### ISSUE TITLE
Autoscaling CoreDNS Nodes on Kubernetes Clusters Through Exposed Metrics

### ISSUE THREAD

**Issue by yongtang** [coredns/coredns#3541]
**Please Note:** This is a tracking issue for [Community Bridge Program](https://github.com/cncf/soc/tree/master/communitybridge). Anyone interested in this implementation should check [link](https://github.com/cncf/soc/tree/master/communitybridge) there.

**Please Note:** This is reserved for Community Bridge Program, **we don't accept PRs outside of Community Bridge Program.**

**Autoscaling CoreDNS Nodes on Kubernetes Clusters Through Exposed Metrics**

- Description: [CoreDNS](https://github.com/coredns/coredns) is Kubernetes' default DNS server for service discovery. As service discovery with DNS is the (often overlooked) critical piece for clusters, it is important to maintain enough capacity for serving the cluster-wide service discovery. CoreDNS already have a `metrics` plugin with metrics information exposed. In this project, the student will work on the following tasks:
  - Write metrics scraping method to extract the metrics from exposed CoreDNS metrics plugin.
  - Develop a method in transforming collected metrics into desired capacity for CoreDNS nodes.
  - Implement a routine to interact with Kubernetes API to autoscaling CoreDNS node with designed capacity.
- Bonus Task: Use machine learning algorithm for transforming metrics to above mentioned designed capacity. We will provide a start algorithm to help student getting started with machine learning, if student is willing to take this bonus task.
- Recommended Skills: Golang and DNS for this project, basic machine learning for bonus task.
- Mentor(s): Yong Tang @yongtang

Implementation: This project will be implemented as a standalone program. It could be placed under coredns org if finished successfully.

**Comment by rajibmitra** [comment]
Interested to complete all the mentioned task as part of community bridge program @yongtang  , could you please assign this to me.

**Comment by yongtang** [comment]
@rajibmitra The application is done through the community bridge program. Please see https://github.com/cncf/soc/tree/master/communitybridge#how-to-apply for details on how to apply.

**Comment by knrt10** [comment]
This looks interesting. @yongtang till now I have run coredns locally and gone through basic plugin tutorial. I am interested in this, can you please guide me through any of the next steps?

**Comment by yongtang** [comment]
@knrt10 you can look into https://github.com/cncf/soc/tree/master/communitybridge#how-to-apply for details on how to apply.

Also, we are still waiting for the update from CNCF on how the slot is allocated (not confirmed yet).

**Comment by knrt10** [comment]
Sure, thanks

**Comment by miekg** [comment]
Sounds like a really nice project. Thanks for doing this @yongtang!

**Comment by chrisohaver** [comment]
Very useful project! Thanks!

> Develop a method in transforming collected metrics into desired capacity for CoreDNS nodes.

I suspect this step is probably more tricky to get right than it might initially seem.
Some incomplete thoughts...

Basing this on total QPS would be tricky to tune, since the QPS ceiling for any given instance is dependent on the compute capacity of the node it is scheduled on.

A mean response latency threshold might be easier to tune. One of the tricks/gotchas here will be _not_ to trigger scaling up based on _upstream_ latency (if an upstream DNS server is slow, it doesnt matter how many CoreDNS instances we spin up, it will still be slow).

Dropped/timed-out requests might be a good measure, but isn't particularly preemptive. I'm not sure how well CoreDNS could really report dropped packets reliably. Also, an upstream could be dropping packets, and we would not want that to trigger a scale up.

One thing to consider would be to send a periodic probe query for a known cluster service (e.g. `kubernetes.default.svc.cluster.local.`) to gauge latency instead of scraping exposed prometheus metrics.

Looking forward to see what comes out of this!

**Comment by miekg** [comment]
The *health* plugin exports `coredns_health_request_duration_seconds` which has been made for just this use case

**Comment by yongtang** [comment]
Thanks for all that applied the project. Due to the limited seat available in this program, we can only confirm one seat. We do want to mention that CoreDNS will continue to participate in Community Bridge and Google-Summer-of-Code, so there are lots of interesting projects for anyone that want to participate.

This issue will be kept open until the program complete. We will share all the ongoing discussions and progress.

Again, thanks all for the participation 👍 !

**Comment by yongtang** [comment]
To share some ideas and hopefully be helpful to the community, here is some thing I found out to quickly get scraping work.

The easiest path so far, is to take a look at https://github.com/prometheus/prom2json. prom2json is a library that do the scraping and convert to json file. We are not dealing with json files but prom2json could be easily imported and used to get scraping. 

Specifically, `prom2json.FetchMetricFamilies` could be reused for our purposes. /cc @knrt10

**Comment by knrt10** [comment]
Thanks, @yongtang , let me look into it

**Comment by knrt10** [comment]
Above referenced PR is as follows:-

The initial plan is to scrape the metrics. We have a new file `scrape.go` inside `plugin/metrics` folder which has a method `ScrapeMetrics` which will help to scrape the metrics and convert it into JSON for the metrics URL we provide.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Autoscaling C..." ]
>Above referenced PR is as follows:-
>
>The initial plan is to scrape the metrics. We have a new file `scrape.go` which has a method `ScrapeMetrics` which will help to scrape the metrics and convert it into JSON for the metrics URL we provide.

2 things about this approach:

1. you can access metrics internally (if you must) - no need to somehow scrape them and
   then convert them into what you need

2. scraping is supported by going to the endpoint and retrieving the metrics - this
   doesn't need new code.


I'm also unclear on what needs to be scraped? Hence my request to have some kind of plan
worked out. What's done with the scraped data? What data do you actually need? Etc. etc.

**Comment by yongtang** [comment]
@knrt10 @miekg Sorry for not adding my comment earlier as I was traveling last week. Some general comment about the scraping. For this project, I think there are several ways to get the metrics, transform to scaling factor (memory/cpu/etc) and feed into k8s:
1) Build this project as a plugin of CoreDNS, and, as Miek mentioned, access metrics internally.
2) Build this project as a standalone program (outside of CoreDNS binary itself), and scraping the metrics from coredns' endpoint directly.
3) You can also setup a prometheus server, and use prometheus to scrap the metrics. Then you can access prometheus server to get the metrics.

There are certainly pros and cons for above mentioned approaches: If you already have a prometheus server setup, then 3) would be easier. However, if you don't already have prometheus server in place, then setting a prometheus server just for the purpose of autoscaling coredns might be an overkill. Thus 1) and 2) might make sense if you want to have a "slim" system for this project.

With respect to 1) and 2), @knrt10 I think it largely depending on how you want to implement the project, either as a coredns plugin, or as a standalone program.

In case of a standalone program, I think having a separate repo would make more sense (could be part of the coredns organization if completed successfully). Otherwise, it could be a coredns plugin.

@knrt10 In any case, I would advice to have a complete working solution first (not a "partial PR").

**Comment by knrt10** [comment]
Sure, let me work on that.

**Comment by SuperQ** [comment]
I'm not sure the approach here makes sense. Autoscaling on Kubernetes typically done via PromQL queries and doesn't need any additional code support from the application.

See:
* https://livewyer.io/blog/2019/05/28/horizontal-pod-autoscaling/
* https://itnext.io/tutorial-auto-scale-your-kubernetes-apps-with-prometheus-and-keda-c6ea460e4642
* https://www.section.io/blog/horizontal-pod-autoscaling-custom-metrics/

**Comment by yongtang** [comment]
@SuperQ As mentioned above (https://github.com/coredns/coredns/issues/3541#issuecomment-570894864 3),  there are some considerations about using prometheus server or not. If there is a prometheus server already setup, then it certainly would be easier for prometheus server to scrap the metrics and wire up with some existing autoscaler. However, if there is no prometheus before hand, then adding a prometheus server for this purpose could be a overkill.

But a bigger question is the "auto-scaling" algorithm itself. As was mentioned in the scope of this project, one goal is to use machine learning algorithm to transform the metrics into autoscaling factors. This is different from a direct autoscaling based on the *current metrics*.

We are looking for potential pre-provisional so that scaling is not based on the existing metrics directly. But based on the prediction of future scaling factor needs according to the past history of the metrics.

For example, if we see a continuous increasing of the memory usage from say, 100MB to 500MB in the past hour, and from the history of past 7 days we know that the same trends lead to an usage of 1GB shortly, then the scaling factor should have 1GB (not current 0.5GB).

This is just a simple example, in reality it would be ideal to take more metrics inputs to see what is the trend over time. Also, will some dns specific metrics (such as number of queries) plays a role in feeding into  seemingly unrelated factors? That is something we want to see as well.

I haven't see an existing autoscaler that allows a specification of a customized machine learning algorithms, thus it makes sense to have an implementation that achieve this goal.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Autoscaling C..." ]
>I haven't see an existing autoscaler that allows a specification of a customized machine learning algorithms, thus it makes sense to have an implementation that achieve this goal.

I thought metrics-server was explicitly designed to handle use-cases like this (but may
be not - worth investigating).

**Comment by yongtang** [comment]
@miekg The metrics-server exposes the usage of CPU/memory from kuberentes' view. However, it does not taking into consideration of DNS (or CoreDNS) specific metrics such as `coredns_health_request_duration_seconds`. In that case, just combining metrics-server with k8s auto-scaling might not achieve the ideal situation.

For example, imagine with 100MB + 0.5CPU coredns could be working correctly but with very long latency.

In that case coredns-specific metrics (such as `coredns_health_request_duration_seconds`) may be needed to get the ideal scaling factor.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Autoscaling C..." ]
>@miekg The metrics-server exposes the usage of CPU/memory from kuberentes' view. However, it does not taking into consideration of DNS (or CoreDNS) specific metrics such as `coredns_health_request_duration_seconds`. In that case, just combining metrics-server with k8s auto-scaling might not achieve the ideal situation.
>
>For example, imagine with 100MB + 0.5CPU coredns could be working correctly but with very long latency.
>
>In that case coredns-specific metrics (such as `coredns_health_request_duration_seconds`) may be needed to get the ideal scaling factor.

Ack and understood. From talking to the devs (a while back) I understood that metrics is
capable (or could be made capable) of doing more with the metrics it gathers.

I'm also not married to k8s - if a k8s agnostic way is found it would be even better.

**Comment by zouyee** [comment]
> I'm not sure the approach here makes sense. Autoscaling on Kubernetes typically done via PromQL queries and doesn't need any additional code support from the application.
> 
> See:
> 
> * https://livewyer.io/blog/2019/05/28/horizontal-pod-autoscaling/
> * https://itnext.io/tutorial-auto-scale-your-kubernetes-apps-with-prometheus-and-keda-c6ea460e4642
> * https://www.section.io/blog/horizontal-pod-autoscaling-custom-metrics/

Kubernetes 1.6 adds support for making use of custom metrics in the Horizontal Pod Autoscaler.
ref https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#support-for-custom-metrics

**Comment by SuperQ** [comment]
Those kinds of decisions can already be achieved via PromQL. There are lots of examples.

Without Prometheus involved, you're inventing a new monitoring system.That sounds out of scope for the CoreDNS project and shouldn't be included here. Any new monitoring system should use the existing metrics interface, which uses the defacto standard Prometheus format used commonly across Kubernetes systems.

**Comment by yongtang** [comment]
With metrics (through prometheus server or directly scraping) and machine learning there are two usage types:
1) For model training, so that the pattern/behavior of the metrics could be analyzed
2) For model inference, where the model obtained in 1) could be used and applied (in realtime) for autoscaling or anomaly detection.

The model training with metrics, is less of a challenge in terms of formats or input sources. This is because most of the time model training are offline. This could be done through:
1) Normalize the features (vector of query latency, query frequency, CPU percentage, memory usage, etc.) into floating points.
2) Build the model with features as input, and generates output. (AutoEncoder could be one possible choices).
   This is normally done with Keras' `model.fit()`

Since training is offline, the input from prometheus server (such as through PromQL for a period of time like past year) could be easily saved in a simple format and use tf.keras (if Keras is used) for training.

Once the training is done, the model will be saved (in h5 or tf format with keras) so that it could be used for inference.

The inference might be challenging because it may need to be done in real time. In this case metrics at each specific time will need to be obtained and passed to Keras' `model.predict` in a continuous fashion. Scraping directly for inference could help avoid intermediate layers and improve response.

But in general, scraping directly or query prometheus server is actually not the focus of this project.  I am less concerned with the scraping vs query prometheus server, as "query prometheus server" technically is a "special" scraping anyway.

In this project the challenge is to integrate metrics with machine learning which are very much different in terms of implementation: metrics collection and processing are mostly in golang while machine learning are mostly in python (to an extent C++). How to build a pipeline so that inference could be done in a continuous pipeline?

I do want to mention that metrics with machine is not something new already: AWS introduced CloudWatch Anomaly Detection which is exactly the metrics + machine learning for anomaly detection.

In this project the focus is metrics + machine learning for autoscaling, but anomaly detection could be equally applied as well.

The only limitation with AWS product (or cloud-vendor provided product) is that the model behind those product are totally black-box: there is no mentioning of how exactly is the model is built and trained?

From that standpoint it makes sense to have an open source implementation to combine machine learning with metrics, so that everything defined (model/training data/etc) is transparent.

This could even be directly done in prometheus implementation (if there is interest from prometheus project maintainers).

The Community Bridge program we are participating here is for CoreDNS only, with a limited time period (ends in March). For that reason we want to limit the scope to be just a CoreDNS related implementation. It could very well becomes a prototype for future prometheus implementation, though.

**Comment by yongtang** [comment]
@knrt10 Any update or I can help clarify some of the edge cases?

One of the bonus task for this project is to add machine learning to the picture. I am working on having an AutoEncoder model as a starter and see if it makes sense. Will help you get started later if you are interested.

In the mean time,  if you can have a basic running pipeline from coredns to Kubernetes autoscaling that would be great to get to the next step. We can always wire up different algorithms or tricks into the mix once a basic pipeline is up.

**Comment by yongtang** [comment]
For anyone that might be interested, I am working on a tutorial of TF+Prometheus+CoreDNS for reading metrics into tf.data.Dataset and used with tf.keras for training a vanilla LSTM model as a starter.

See tensorflow/io#774 for the PR. You can also check the tutorial from Google Colab in https://colab.research.google.com/github/yongtang/io/blob/prometheus/docs/tutorials/prometheus.ipynb

**Comment by yongtang** [comment]
This project is closed and is moved to coredns/rfc.
